src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,130 @@
+import java.io.FileReader;
+import java.io.FileWriter;
+
+import java.util.ArrayList;
+
+import java.time.LocalDate;
+
+import java.io.IOException;
+import java.io.FileNotFoundException;
+
+import commands.*;
+import tasks.*;
+import exceptions.*;
+
+/**
+ * This class is responsible for reading and writing to the file for a Duke program.
+ *
+ * On initialization, it will contain a File object to which it will read and write to.
+ * If the file path specified does not exist, it will instead create a new clean file
+ * for writing to.
+ */
+public class Storage {
+
+    private File taskFile;
+
+    Storage(String filePath) {
+        String localDir = System.getProperty("user.dir");
+        this.taskFile = new File(localDir + File.separator + filePath);
+        if (!taskFile.exists()) {
+            taskFile.getParentFile().mkdirs();
+            try {
+                taskFile.createNewFile();
+            } catch (IOException e) {
+                System.out.println("Failed to create new file");
+            }
+        }
+    }
+
+    /**
+     * Takes in a line from the task file saved in disk and processes it
+     * by adding it to the provided ArrayList of Tasks.
+     *
+     * @param taskLine A line from the file that is being read from.
+     * @param taskList The ArrayList which will contain the Tasks in the file.
+     * @throws DukeException If the letter in each line does not match any of the task types.
+     */
+    private void process(String taskLine, ArrayList<Task> taskList) throws DukeException {
+        String[] parsedLine = taskLine.split(" \\| ", 3);
+        String command = parsedLine[0];
+        Boolean isDone = parsedLine[1].equals("1");
+        String next = parsedLine[2];
+        if (command.equals("T")) {
+            Todo todo = new Todo(next, isDone);
+            taskList.add(todo);
+        } else if (command.equals("D")) {
+            String[] details = next.split(" \\| ", 2);
+            String desc = details[0];
+            LocalDate dueDate = LocalDate.parse(details[1]);
+            Deadline dl = new Deadline(desc, isDone, dueDate);
+            taskList.add(dl);
+        } else if (command.equals("E")) {
+            String[] details = next.split(" \\| ", 2);
+            String desc = details[0];
+            String time = details[1];
+            Event e = new Event(desc, isDone, time);
+            taskList.add(e);
+        } else {
+            throw new DukeException();
+        }
+    }
+
+    /**
+     * Reads the file of the current Storage object and stores the Tasks
+     * saved in the file into a ArrayList of Tasks.
+     *
+     * @return The ArrayList of Tasks read from the file.
+     */
+    public ArrayList<Task> readFile() {
+        try (BufferedReader br = new BufferedReader(new FileReader(this.taskFile))) {
+            ArrayList<Task> taskList = new ArrayList<>();
+            String line;
+            while ((line = br.readLine()) != null && !line.equals("")) {
```

Comment:
Not sure if there is a clearer way to express this? Not sure if it's considered a complicated expression (textbook: page 57-58)